students/brian_minsk/lesson05/Assignments/mailroom3.py

```diff
@@ -0,0 +1,306 @@
+
+    Keyword arguments:
+    donor_name = name of the donor
+    donation_amount = donation amount as a float
+    """
+    donor = get_donor(donor_name)
+    message_line1 = "\nDear {first_name} {last_name},\n".format(**donor)
+    # Note: I tried to split the following so the line wouldn't be too long
+    # using string continuation like so:
+    #       message_rest = "     Thank you very much for your generous donation of "
+    #       "${:.2f}.\n     WINNING!\n".format(float(donation_amount))
+    # but the output was different. I don't understand.
+    message_rest = "     Thank you very much for your generous donation of ${:.2f}.\n     WINNING!\n".format(float(donation_amount))
+    return message_line1 + message_rest
+
+
+def get_donor(donor_name):
+    """ Return a donor from the donor_db dict if one exists.
+
+    Keyword arguments:
+    donor_name - String with a name, assumed to be of the form
+    "first_name last_name"
+    """
+    donor_id = is_existing_donor(donor_name)
+    if donor_id:
+        return donor_db[donor_id]
+    else:
+        return None
+
+
+def create_report():
+    """ Print a list of your donors, sorted by total historical donation amount.
+    Include Donor Name, total donated, number of donations, and average
+    donation amount as values in each row. The end result should be tabular
+    (values in each column should align with those above and below) and look
+    something like this:
+
+    Donor Name                | Total Given | Num Gifts | Average Gift
+    ------------------------------------------------------------------
+    William Gates, III         $  653784.49           2  $   326892.24
+    Mark Zuckerberg            $   16396.10           3  $     5465.37
+    Jeff Bezos                 $     877.33           1  $      877.33
+    Paul Allen                 $     708.42           3  $      236.14
+    """
+    try:
+        report_rows = create_report_rows()
+    except AssertionError:
+        print("Report creation failed")
+        return
+
+    # print header
+    print("{:<26}| {:<11}| {:<10}| {:<12}".format("Donor Name", "Total Given",
+                                                  "Num Gifts", "Average Gift"))
+    # print header line
+    for __ in range(0, 65):
+        print("-", end="")
+    print("")  # get a new line
+    # print rows
+    print("\n".join(report_rows))
+    print("")
+
+
+def create_report_rows():
+    """ Create the rows (as described in create_report()) with the data from
+    the donor db, sorted by total donation amount (computed by the sort_key
+    function). Return a list with each row as a list item.
+    """
+    sorted_donor_keys = sorted(donor_db, key=sort_key, reverse=True)
+    report_rows = [create_single_report_row(donor_db[donor_key])
+                   for donor_key in sorted_donor_keys]
+    try:
+        assert len(report_rows) == len(donor_db)
+    except AssertionError:
+        raise AssertionError
```

Comment:
why catch if you're just going to raise the exception anyway?

Reply:
I'm a little confused by this. Can you please elaborate a little?

Reply:
:ok_hand: Oh, I get it now. I don't handle the exception other than raising it so it is useless ... 
Thanks for the comment!  :cowboy_hat_face:

Reply:
cool 👍